### PR TITLE
[cm] Custom Neutone exceptions

### PR DESCRIPTION
**Pull request by christhetree**
This PR implements Neutone exceptions which are used when other exceptions have cryptic error messages that should be elaborated on.

Currently takes advantage of the `raise from` functionality introduced in python 3, let me know if you think there's a nicer way to do this.

### CODE REVIEW

File: neutone_sdk/utils.py [hunk 1]
```diff
@@ -112,7 +112,7 @@ def save_neutone_model(
 
     sqw = SampleQueueWrapper(model)
 
-    with tr.no_grad():
+    with tr.inference_mode():
```
Comment (bogdanteleaga) [inline neutone_sdk/utils.py:115]: Do we need this `inference_mode` block here?

File: neutone_sdk/benchmark.py [hunk 2]
```diff
@@ -212,34 +216,36 @@ def profile_sqw(
         sqw.prepare_for_inference()
     if convert_to_torchscript:
         log.info("Converting to TorchScript")
-        with torch.no_grad():
+        with torch.inference_mode():
```
Comment (bogdanteleaga) [inline neutone_sdk/benchmark.py:219]: Also similar for this one, do we actually need it?

File: neutone_sdk/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        """
+        if (isinstance(orig_exception, self.trigger_type)
+                and self.trigger_str in str(orig_exception)):
+            raise self from orig_exception
+        else:
+            raise orig_exception
+
+
+# TODO(cm): constant for now, but if we need more of these we could use a factory method
+INFERENCE_MODE_EXCEPTION = NeutoneException(
```
Comment (bogdanteleaga) [inline neutone_sdk/exceptions.py:40]: Could this be a list of possible exceptions that you iterate through it in the try/catch block?

File: neutone_sdk/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    message="""
+    Your model does not support inference mode. Ensure you are not calling forward on
+    your model before wrapping it or saving it using `save_neutone_model()`. Also, try 
+    to make sure that you are not creating new tensors in the forward call of your
+    model, instead pre-allocate them in the constructor. If these suggestions fail, try 
+    creating and saving your model entirely inside of a `with torch.inference_mode():` 
+    block.
+    """,
+    trigger_type=RuntimeError,
+    trigger_str="Inference tensors cannot be saved for backward."
```
Comment (bogdanteleaga) [inline neutone_sdk/exceptions.py:50]: I am a bit worried about:
- Is this specific enough to not accidentally catch other exceptions?
- We need to test it with different pytorch versions and see if we get the same error string.

File: neutone_sdk/benchmark.py [hunk 5]
```diff
@@ -212,34 +216,36 @@ def profile_sqw(
...
-        log.info(
-            prof.key_averages().table(sort_by="self_cpu_memory_usage", row_limit=10)
-        )
-        log.info("Displaying Grouped CPU Memory Usage")
-        log.info(
-            prof.key_averages(group_by_stack_n=5).table(
-                sort_by="self_cpu_memory_usage", row_limit=5
+            log.info("Displaying Total CPU Time")
+            log.info(prof.key_averages().table(sort_by="cpu_time_total", row_limit=10))
+            # log.info(prof.key_averages(group_by_stack_n=5).table(sort_by="cpu_time_total", row_limit=10))
```
Comment (bogdanteleaga) [inline neutone_sdk/benchmark.py:236]: Can you remove this?

**Review by bogdanteleaga — COMMENTED** [review]
Looks good overall, thank you!
Left a couple of comments. I like the `raise from` approach, looks like we could also do a `raise from None` to suppress the original message. I think that's fine in this case, but maybe too much in general?

File: neutone_sdk/exceptions.py (hunk 4 above)
Comment (christhetree) [reply neutone_sdk/exceptions.py:50]: Agreed, I used only a snippet of the text, but it's probably better to change it to use almost the entire text. I think it would also make sense to change this to a list of strings that can each trigger the exception such that if the text is slightly different in older pytorch versions we can simply add those messages to the list.
